src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/NuGetDependenciesSubTreeProvider.cs

```diff
@@ -216,96 +227,259 @@ public override IDependencyNode GetDependencyNode(DependencyNodeId id)
+                    if (itemNode != null)
                     {
-                        CurrentSnapshot.RemoveDependency(removedItemSpec);
+                        dependenciesChange.RemovedNodes.Add(itemNode);
+                    }
+                }
 
-                        var itemNode = RootNode.Children.FirstOrDefault(
-                                        x => x.Id.ItemSpec.Equals(removedItemSpec, StringComparison.OrdinalIgnoreCase));
-                        if (itemNode != null)
-                        {
-                            dependenciesChange.RemovedNodes.Add(itemNode);
-                        }
+                // update
+                foreach (var resolvedMetadata in resolvedChanges.UpdatedNodes)
+                {
+                    var itemNode = rootNodes.FirstOrDefault(
+                                    x => x.Id.ItemSpec.Equals(resolvedMetadata.ItemSpec, StringComparison.OrdinalIgnoreCase));
+                    if (itemNode != null)
+                    {
+                        itemNode = CreateDependencyNode(resolvedMetadata, topLevel: true);
+                        dependenciesChange.UpdatedNodes.Add(itemNode);
                     }
-                    
-                    foreach (string changedItemSpec in change.Difference.ChangedItems)
+
+                    var unresolvedMatch = unresolvedChanges.UpdatedNodes.FirstOrDefault(x => x.ItemSpec.Equals(resolvedMetadata.Name));
+                    if (unresolvedMatch != null)
                     {
-                        var properties = GetProjectItemProperties(change.After, changedItemSpec);
-                        if (properties == null)
-                        {
-                            continue;
-                        }
+                        unresolvedChanges.UpdatedNodes.Remove(unresolvedMatch);
+                    }
+                }
 
-                        CurrentSnapshot.UpdateDependency(changedItemSpec, properties);
+                foreach (var unresolvedMetadata in unresolvedChanges.UpdatedNodes)
+                {
+                    var itemNode = rootNodes.FirstOrDefault(
+                                    x => x.Id.ItemSpec.Equals(unresolvedMetadata.ItemSpec, StringComparison.OrdinalIgnoreCase));
+                    if (itemNode != null)
+                    {
+                        dependenciesChange.UpdatedNodes.Add(itemNode);
+                    }
+                }
 
-                        var itemNode = RootNode.Children.FirstOrDefault(
-                                        x => x.Id.ItemSpec.Equals(changedItemSpec, StringComparison.OrdinalIgnoreCase));
-                        if (itemNode != null)
-                        {
-                            dependenciesChange.UpdatedNodes.Add(itemNode);
-                        }
+                // add
+                foreach (var resolvedMetadata in resolvedChanges.AddedNodes)
```

Comment:
Actually i was testing it more and there is a scenario when your case is valid: when project is loading and there no root nodes added yet and we have unresolved and resolved nodes sent in one snapshot together


---
In reply to: [85950255](https://github.com/dotnet/roslyn-project-system/pull/675#discussion_r85950255) [](ancestors = 85950255,85852326)